staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go

```diff
@@ -356,3 +368,61 @@ func tooManyRequests(req *http.Request, w http.ResponseWriter, retryAfter string
 	w.Header().Set("Retry-After", retryAfter)
 	http.Error(w, "Too many requests, please try again later.", http.StatusTooManyRequests)
 }
+
+func newRequestWaitContextFactory(clock utilsclock.PassiveClock, defaultRequestWaitLimit time.Duration) *requestWaitContextFactory {
+	return &requestWaitContextFactory{
+		clock:                   clock,
+		defaultRequestWaitLimit: defaultRequestWaitLimit,
+		newCtxFn:                context.WithDeadline,
+	}
+}
+
+type requestWaitContextFactory struct {
+	clock utilsclock.PassiveClock
+
+	// the default wait duration that is used if the request context does
+	// not have any deadline, the wait limited used for:
+	// (a) initialization of a watch or
+	// (b) a request whose context has no deadline
+	defaultRequestWaitLimit time.Duration
+
+	// comes handy to write unit test
+	newCtxFn func(parent context.Context, at time.Time) (context.Context, context.CancelFunc)
+}
+
+// GetRequestWaitContext returns a new context with a deadline of how
+// long the request is allowed to wait before it is removed from its
+// queue and rejected.
+// The context.CancelFunc returned must never be nil and the caller is
+// responsible for calling the CancelFunc function for cleanup.
+//   - ctx: the context associated with the request (it may or may
+//     not have a deadline).
+//   - isWatch: true if the given request is a WATCH, otherwise false
+func (rwcf *requestWaitContextFactory) GetRequestWaitContext(ctx context.Context, isWatch bool) (context.Context, context.CancelFunc) {
+	if ctx.Err() != nil {
+		return ctx, func() {}
+	}
+
+	reqArrivedAt := rwcf.clock.Now()
+	if reqReceivedTimestamp, ok := apirequest.ReceivedTimestampFrom(ctx); ok {
+		reqArrivedAt = reqReceivedTimestamp
+	}
+
+	if isWatch {
+		return rwcf.newCtxFn(ctx, reqArrivedAt.Add(rwcf.defaultRequestWaitLimit))
```

Comment:
Why do we want to use different deadline for watch requests?

Reply:
For a WATCH request the context's deadline is normally minutes away, we do not want a quarter of _that_.

Reply:
Why?
For watch requests we (by default) use a timeout of random from [5m, 10m] interval. So a quarter from that would be [1m15s, 2m30s].
Additionally, watch by definition is "eventually consistent", so I'm not sure I see a strong reason why we can't use that and instead of that use 1m...

Reply:
For a WATCH request, the current merged behavior is a wait limit of 1/4 of the server's normal request time limit (the thing that defaults to 1 minute).  In order to preserve the current behavior, the revised code has to handle WATCH requests specially, because their Context's deadline does not reflect that defaults-to-1-minute limit.

Reply:
> In order to preserve the current behavior, the revised code has to handle WATCH requests specially, because their Context's deadline does not reflect that defaults-to-1-minute limit.

I don't think it's the case, today we don't set any deadline if the request is a WATCH or a long running request in server filter. The deadline for a watch gets set below, which is after apf filter.

https://github.com/kubernetes/kubernetes/blob/1c651cf7398f612124297dbb47f30a6825337759/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go#L47-L60

Maybe we should move it inside the filter?
https://github.com/kubernetes/kubernetes/blob/1c651cf7398f612124297dbb47f30a6825337759/staging/src/k8s.io/apiserver/pkg/endpoints/filters/request_deadline.go#L54


so for a WATCH, this PR is going to set a `15s` deadline
for now, I can add a unit test that fails if the request is a WATCH and has a deadline so that if we ever set deadline for a WATCH inside server filter then this test will fail and give us a chance to revise the wait limit for WATCH then.

Reply:
@MikeSpreitzer @wojtek-t  does the following sound reasonable?

```
switch {
case hasDeadline:
     waitLimit = min(ctx.Deadline() / 4, 1m)
default:
     waitLimit = defaultWaitLimit (1/4th of configured request-timeout)
}
```
with this, we don't need to special case for a watch request

Reply:
That sounds reasonable to me.

Reply:
I just wired the PR with this behavior,  please take a look when you have time.